addons/stock/models/stock_quant.py

```diff
@@ -759,19 +773,48 @@ def _get_available_quantity(self, product_id, location_id, lot_id=None, package_
             else:
                 return sum([available_quantity for available_quantity in availaible_quantities.values() if float_compare(available_quantity, 0, precision_rounding=rounding) > 0])
 
-    def _get_reserve_quantity(self, product_id, location_id, quantity, lot_id=None, package_id=None, owner_id=None, strict=False):
+    def _get_reserve_quantity(self, product_id, location_id, quantity, product_packaging_id=None, uom_id=None, lot_id=None, package_id=None, owner_id=None, strict=False, only_available=False):
         """ Get the quantity available to reserve for the set of quants
         sharing the combination of `product_id, location_id` if `strict` is set to False or sharing
-        the *exact same characteristics* otherwise. Typically, this method is called before the
-        `stock.move.line` creation to know the reserved_qty that could be use. It's also call by
-        `_update_reserve_quantity` to find the quant to reserve.
+        the *exact same characteristics* otherwise. If no quants are in self, `_gather` will do a search to fetch the quants
+        If `only_available` is set, only available qty is reserved.
+        Typically, this method is called before the `stock.move.line` creation to know the reserved_qty that could be use.
+        It's also called by `_update_reserve_quantity` to find the quant to reserve.
 
         :return: a list of tuples (quant, quantity_reserved) showing on which quant the reservation
             could be done and how much the system is able to reserve on it
         """
         self = self.sudo()
         rounding = product_id.uom_id.rounding
-        quants = self or self._gather(product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict, qty=quantity)
+
+        quants = self._gather(product_id, location_id, lot_id=lot_id, package_id=package_id, owner_id=owner_id, strict=strict, qty=quantity)
+
+        # avoid quants with negative qty to not lower available_qty
+        available_quantity = sum(quants.filtered(lambda q: q.available_quantity > 0).mapped('available_quantity'))
+
+        # do full packaging reservation when it's needed
+        if product_packaging_id and product_id.product_tmpl_id.categ_id.packaging_reserve_method == "full":
+            available_quantity = product_packaging_id._check_qty(available_quantity, product_id.uom_id, "DOWN")
+
+        if only_available:
+            quantity = min(quantity, available_quantity)
```

Comment:
All the computation with available_quantity will be false if there was a quant with negative quantity due to concurrency.